File: modules/agent/src/services/organization_service.py
"""Service layer for organization business logic."""

import logging
from typing import Dict, List, Optional, Any

from fastapi import HTTPException

from repositories.contact_repository import (
    create_contact as create_contact_repo,
    update_contact as update_contact_repo,
    delete_contact as delete_contact_repo,
    find_contact_by_id,
    find_contacts_by_account,
    link_contact_to_account,
    find_contact_account_link,
    clear_primary_contacts_for_account,
)
from repositories.organization_repository import (
    find_all_organizations,
    find_organization_by_id,
    create_organization as create_organization_repo,
    update_organization as update_organization_repo,
    delete_organization as delete_organization_repo,
    search_organizations as search_organizations_repo,
    get_organization_account_id,
    create_account_for_organization,
    update_account_industries,
    update_account_projects,
    find_organization_relationships as find_organization_relationships_repo,
    find_organization_relationship,
    create_organization_relationship as create_organization_relationship_repo,
    find_organization_relationship_by_id,
    delete_organization_relationship_by_id,
    OrganizationCreate,
    OrganizationUpdate,
    OrgContactCreate,
    OrgContactUpdate,
    OrgTechnologyCreate,
    FundingRoundCreate,
    SignalCreate,
)
from repositories.individual_repository import get_individual_account_id

# Re-export Pydantic models for controllers
__all__ = [
    "OrganizationCreate",
    "OrganizationUpdate",
    "OrgContactCreate",
    "OrgContactUpdate",
    "OrgTechnologyCreate",
    "FundingRoundCreate",
    "SignalCreate",
]
from repositories.technology_repository import (
    find_organization_technologies,
    add_organization_technology,
    remove_organization_technology,
)
from repositories.funding_round_repository import (
    find_funding_rounds_by_org,
    create_funding_round as create_funding_round_repo,
    delete_funding_round as delete_funding_round_repo,
    find_funding_round_by_id,
)
from repositories.signal_repository import (
    find_signals_by_account,
    create_signal as create_signal_repo,
    delete_signal as delete_signal_repo,
    find_signal_by_id,
)

logger = logging.getLogger(__name__)


async def get_organizations_paginated(
    tenant_id: Optional[int],
    page: int,
    page_size: int,
    order_by: str,
    order: str,
    search: Optional[str] = None,
):
    """Get organizations with pagination."""
    return await find_all_organizations(
        tenant_id=tenant_id,
        page=page,
        page_size=page_size,
        order_by=order_by,
        order=order,
        search=search,
    )


async def search_organizations(query: str, tenant_id: Optional[int]):
    """Search organizations by name."""
    return await search_organizations_repo(query, tenant_id=tenant_id)


async def get_organization(org_id: int):
    """Get organization by ID with contacts."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    contacts = await find_contacts_by_account(org.account_id) if org.account_id else []
    return org, contacts


async def create_organization(
    org_data: Dict[str, Any],
    tenant_id: Optional[int],
    industry_ids: Optional[List[int]],
    project_ids: Optional[List[int]],
):
    """Create organization with associated account, industries, and projects."""
    # Create an Account for the Organization if not provided
    if not org_data.get("account_id"):
        account_id = await create_account_for_organization(
            tenant_id=tenant_id,
            account_name=org_data.get("name", ""),
            created_by=org_data.get("created_by"),
            updated_by=org_data.get("updated_by"),
            industry_ids=industry_ids,
            project_ids=project_ids,
        )
        org_data["account_id"] = account_id

    try:
        return await create_organization_repo(org_data)
    except Exception as e:
        if "idx_organization_name_lower" in str(e):
            raise HTTPException(status_code=400, detail="An organization with this name already exists")
        raise


async def update_organization(
    org_id: int,
    org_data: Dict[str, Any],
    industry_ids: Optional[List[int]],
    project_ids: Optional[List[int]],
    project_ids_provided: bool,
):
    """Update organization with industries and projects."""
    org = await update_organization_repo(org_id, org_data)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")

    # Update industries if provided
    if industry_ids is not None and org.account_id:
        await update_account_industries(org.account_id, industry_ids)

    # Update projects if explicitly provided
    if project_ids_provided and org.account_id:
        await update_account_projects(org.account_id, project_ids)

    # Re-fetch to get updated data
    return await find_organization_by_id(org_id)


async def delete_organization(org_id: int):
    """Delete an organization."""
    success = await delete_organization_repo(org_id)
    if not success:
        raise HTTPException(status_code=404, detail="Organization not found")
    return True


# Contact management

async def get_organization_contacts(org_id: int):
    """Get contacts for an organization."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    return await find_contacts_by_account(org.account_id) if org.account_id else []


async def create_organization_contact(
    org_id: int,
    contact_data: Dict[str, Any],
    user_id: Optional[int],
):
    """Create a contact for an organization."""
    org = await find_organization_by_id(org_id)
    if not org or not org.account_id:
        raise HTTPException(status_code=404, detail="Organization not found")

    # Create contact via repository
    contact = await create_contact_repo({
        "first_name": contact_data.get("first_name"),
        "last_name": contact_data.get("last_name"),
        "title": contact_data.get("title"),
        "department": contact_data.get("department"),
        "role": contact_data.get("role"),
        "email": contact_data.get("email"),
        "phone": contact_data.get("phone"),
        "is_primary": contact_data.get("is_primary", False),
        "notes": contact_data.get("notes"),
        "created_by": user_id,
        "updated_by": user_id,
    })

    # Link to organization's account
    await link_contact_to_account(contact.id, org.account_id, contact_data.get("is_primary", False))

    # Link to another organization's account if provided (partner relationship)
    if contact_data.get("linked_organization_id"):
        partner_account_id = await get_organization_account_id(contact_data["linked_organization_id"])
        if partner_account_id:
            await link_contact_to_account(contact.id, partner_account_id, False)

    # Link to individual's account if provided (for relationships)
    if contact_data.get("individual_id"):
        ind_account_id = await get_individual_account_id(contact_data["individual_id"])
        if ind_account_id:
            await link_contact_to_account(contact.id, ind_account_id, False)

    return await find_contact_by_id(contact.id)


async def update_organization_contact(
    org_id: int,
    contact_id: int,
    contact_data: Dict[str, Any],
    user_id: Optional[int],
):
    """Update a contact for an organization."""
    org = await find_organization_by_id(org_id)
    if not org or not org.account_id:
        raise HTTPException(status_code=404, detail="Organization not found")

    # Check contact is linked to this organization's account
    if not await find_contact_account_link(contact_id, org.account_id):
        raise HTTPException(status_code=404, detail="Contact not found for this organization")

    # Build update data
    update_data = {}
    for field in ["first_name", "last_name", "title", "department", "role", "email", "phone", "notes"]:
        if contact_data.get(field) is not None:
            update_data[field] = contact_data[field]

    # Handle is_primary specially
    if contact_data.get("is_primary"):
        await clear_primary_contacts_for_account(org.account_id, contact_id)
        update_data["is_primary"] = True
    elif "is_primary" in contact_data and contact_data["is_primary"] is not None:
        update_data["is_primary"] = contact_data["is_primary"]

    update_data["updated_by"] = user_id
    return await update_contact_repo(contact_id, update_data)


async def delete_organization_contact(org_id: int, contact_id: int):
    """Delete a contact from an organization."""
    org = await find_organization_by_id(org_id)
    if not org or not org.account_id:
        raise HTTPException(status_code=404, detail="Organization not found")

    # Check contact is linked to this organization's account
    if not await find_contact_account_link(contact_id, org.account_id):
        raise HTTPException(status_code=404, detail="Contact not found for this organization")

    success = await delete_contact_repo(contact_id)
    if not success:
        raise HTTPException(status_code=404, detail="Contact not found")
    return True


# Technology management

async def get_technologies(org_id: int):
    """Get technologies for an organization."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    return await find_organization_technologies(org_id)


async def add_technology(org_id: int, tech_id: int, source: Optional[str], confidence: Optional[float]):
    """Add a technology to an organization."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    return await add_organization_technology(
        org_id=org_id,
        tech_id=tech_id,
        source=source,
        confidence=confidence
    )


async def remove_technology(org_id: int, technology_id: int):
    """Remove a technology from an organization."""
    success = await remove_organization_technology(org_id, technology_id)
    if not success:
        raise HTTPException(status_code=404, detail="Technology not found for this organization")
    return True


# Funding round management

async def get_funding_rounds(org_id: int):
    """Get funding rounds for an organization."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    return await find_funding_rounds_by_org(org_id)


async def create_funding_round(org_id: int, round_data: Dict[str, Any]):
    """Create a funding round for an organization."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    return await create_funding_round_repo(
        org_id=org_id,
        round_type=round_data.get("round_type"),
        amount=round_data.get("amount"),
        announced_date=round_data.get("announced_date"),
        lead_investor=round_data.get("lead_investor"),
        source_url=round_data.get("source_url")
    )


async def delete_funding_round(org_id: int, round_id: int):
    """Delete a funding round."""
    funding_round = await find_funding_round_by_id(round_id)
    if not funding_round or funding_round.organization_id != org_id:
        raise HTTPException(status_code=404, detail="Funding round not found")
    await delete_funding_round_repo(round_id)
    return True


# Signal management

async def get_signals(org_id: int, signal_type: Optional[str], limit: int):
    """Get signals for an organization (via its account)."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    if not org.account_id:
        return []
    return await find_signals_by_account(org.account_id, signal_type=signal_type, limit=limit)


async def create_signal(org_id: int, signal_data: Dict[str, Any]):
    """Create a signal for an organization (via its account)."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    if not org.account_id:
        raise HTTPException(status_code=400, detail="Organization has no account")
    return await create_signal_repo(
        account_id=org.account_id,
        signal_type=signal_data.get("signal_type"),
        headline=signal_data.get("headline"),
        description=signal_data.get("description"),
        signal_date=signal_data.get("signal_date"),
        source=signal_data.get("source"),
        source_url=signal_data.get("source_url"),
        sentiment=signal_data.get("sentiment"),
        relevance_score=signal_data.get("relevance_score")
    )


async def delete_signal(org_id: int, signal_id: int):
    """Delete a signal."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    signal = await find_signal_by_id(signal_id)
    if not signal or signal.account_id != org.account_id:
        raise HTTPException(status_code=404, detail="Signal not found")
    await delete_signal_repo(signal_id)
    return True


# Organization-to-Organization Relationship management

async def get_organization_relationships(org_id: int):
    """Get all relationships for an organization (both directions)."""
    org = await find_organization_by_id(org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")

    return await find_organization_relationships_repo(org_id)


async def create_organization_relationship(
    from_org_id: int,
    to_org_id: int,
    relationship_type: str = None,
    notes: str = None,
):
    """Create a relationship between two organizations."""
    if from_org_id == to_org_id:
        raise HTTPException(status_code=400, detail="Cannot create relationship with self")

    from_org = await find_organization_by_id(from_org_id)
    if not from_org:
        raise HTTPException(status_code=404, detail="Source organization not found")

    to_org = await find_organization_by_id(to_org_id)
    if not to_org:
        raise HTTPException(status_code=404, detail="Target organization not found")

    # Check if relationship already exists
    if await find_organization_relationship(from_org_id, to_org_id):
        raise HTTPException(status_code=400, detail="Relationship already exists")

    return await create_organization_relationship_repo(
        from_org_id=from_org_id,
        to_org_id=to_org_id,
        relationship_type=relationship_type,
        notes=notes,
    )


async def delete_organization_relationship(from_org_id: int, relationship_id: int):
    """Delete a relationship."""
    relationship = await find_organization_relationship_by_id(relationship_id, from_org_id)
    if not relationship:
        raise HTTPException(status_code=404, detail="Relationship not found")

    await delete_organization_relationship_by_id(relationship_id)
    return True

File: modules/agent/internal/services/organization_service.go
package services

import (
	"errors"

	"github.com/shopspring/decimal"

	"github.com/pina-colada-co/agent-go/internal/lib"
	"github.com/pina-colada-co/agent-go/internal/models"
	"github.com/pina-colada-co/agent-go/internal/repositories"
	"github.com/pina-colada-co/agent-go/internal/schemas"
	"github.com/pina-colada-co/agent-go/internal/serializers"
)

// ErrOrganizationNotFound is returned when an organization is not found
var ErrOrganizationNotFound = errors.New("organization not found")

// OrganizationService handles organization business logic
type OrganizationService struct {
	orgRepo *repositories.OrganizationRepository
}

// NewOrganizationService creates a new organization service
func NewOrganizationService(orgRepo *repositories.OrganizationRepository) *OrganizationService {
	return &OrganizationService{orgRepo: orgRepo}
}

// GetOrganizations returns paginated organizations
func (s *OrganizationService) GetOrganizations(page, pageSize int, orderBy, order, search string, tenantID *int64) (*serializers.PagedResponse, error) {
	params := repositories.NewPaginationParams(page, pageSize, orderBy, order)
	params.Search = search

	result, err := s.orgRepo.FindAll(params, tenantID)
	if err != nil {
		return nil, err
	}

	items := make([]serializers.OrganizationListResponse, len(result.Items))
	for i := range result.Items {
		items[i] = serializers.OrganizationToListResponse(&result.Items[i])
	}

	resp := serializers.NewPagedResponse(items, result.TotalCount, result.Page, result.PageSize, result.TotalPages)
	return &resp, nil
}

// GetOrganization returns an organization by ID
func (s *OrganizationService) GetOrganization(id int64) (*serializers.OrganizationDetailResponse, error) {
	org, err := s.orgRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if org == nil {
		return nil, ErrOrganizationNotFound
	}

	resp := serializers.OrganizationToDetailResponse(org)
	return &resp, nil
}

// SearchOrganizations searches organizations by name
func (s *OrganizationService) SearchOrganizations(query string, tenantID *int64, limit int) ([]serializers.OrganizationBrief, error) {
	if limit <= 0 {
		limit = 10
	}

	orgs, err := s.orgRepo.FindByName(query, tenantID, limit)
	if err != nil {
		return nil, err
	}

	results := make([]serializers.OrganizationBrief, len(orgs))
	for i := range orgs {
		results[i] = serializers.OrganizationBrief{
			ID:   orgs[i].ID,
			Name: orgs[i].Name,
		}
	}

	return results, nil
}

// DeleteOrganization deletes an organization
func (s *OrganizationService) DeleteOrganization(id int64) error {
	org, err := s.orgRepo.FindByID(id)
	if err != nil {
		return err
	}
	if org == nil {
		return ErrOrganizationNotFound
	}

	return s.orgRepo.Delete(id)
}

// UpdateOrganization updates an organization
func (s *OrganizationService) UpdateOrganization(id int64, input schemas.OrganizationUpdate, userID int64) (*serializers.OrganizationDetailResponse, error) {
	org, err := s.orgRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if org == nil {
		return nil, ErrOrganizationNotFound
	}

	if err := s.applyOrgUpdates(org, input); err != nil {
		return nil, err
	}

	if err := s.updateOrgIndustries(org.AccountID, input.IndustryIDs); err != nil {
		return nil, err
	}

	return s.GetOrganization(id)
}

func (s *OrganizationService) applyOrgUpdates(org *models.Organization, input schemas.OrganizationUpdate) error {
	updates := buildOrgUpdates(input)
	if len(updates) == 0 {
		return nil
	}
	return s.orgRepo.Update(org, updates)
}

func (s *OrganizationService) updateOrgIndustries(accountID *int64, industryIDs []int64) error {
	if industryIDs == nil || accountID == nil {
		return nil
	}
	return s.orgRepo.UpdateAccountIndustries(*accountID, industryIDs)
}

func buildOrgUpdates(input schemas.OrganizationUpdate) map[string]interface{} {
	updates := make(map[string]interface{})
	if input.Name != nil {
		updates["name"] = *input.Name
	}
	if input.Website != nil {
		updates["website"] = *input.Website
	}
	if input.Phone != nil {
		updates["phone"] = *input.Phone
	}
	if input.EmployeeCount != nil {
		updates["employee_count"] = *input.EmployeeCount
	}
	if input.Description != nil {
		updates["description"] = *input.Description
	}
	if input.LinkedInURL != nil {
		updates["linkedin_url"] = *input.LinkedInURL
	}
	if input.FoundingYear != nil {
		updates["founded"] = *input.FoundingYear
	}
	if input.HeadquartersCity != nil {
		updates["headquarters"] = *input.HeadquartersCity
	}
	return updates
}

// AddContactToOrganization adds a contact to an organization
func (s *OrganizationService) AddContactToOrganization(orgID int64, input schemas.OrgContactCreate, userID int64, tenantID int64) (*serializers.ContactResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil || org.AccountID == nil {
		return nil, ErrOrganizationNotFound
	}

	repoInput := repositories.OrgContactInput{
		FirstName: input.FirstName,
		LastName:  input.LastName,
		Email:     input.Email,
		Phone:     input.Phone,
		Title:     input.Title,
		IsPrimary: input.IsPrimary,
	}

	contact, err := s.orgRepo.CreateContactForAccount(*org.AccountID, repoInput, userID)
	if err != nil {
		return nil, err
	}

	return &serializers.ContactResponse{
		ID:        contact.ID,
		FirstName: contact.FirstName,
		LastName:  contact.LastName,
		Email:     contact.Email,
		Phone:     contact.Phone,
		Title:     contact.Title,
	}, nil
}

// GetOrganizationContacts returns contacts for an organization
func (s *OrganizationService) GetOrganizationContacts(orgID int64) ([]serializers.ContactResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil || org.AccountID == nil {
		return nil, ErrOrganizationNotFound
	}

	contacts, err := s.orgRepo.GetContactsForAccount(*org.AccountID)
	if err != nil {
		return nil, err
	}

	results := make([]serializers.ContactResponse, len(contacts))
	for i, c := range contacts {
		results[i] = serializers.ContactResponse{
			ID:        c.ID,
			FirstName: c.FirstName,
			LastName:  c.LastName,
			Email:     c.Email,
			Phone:     c.Phone,
			Title:     c.Title,
		}
	}
	return results, nil
}

// UpdateOrganizationContact updates a contact for an organization
func (s *OrganizationService) UpdateOrganizationContact(orgID int64, contactID int64, input schemas.OrgContactUpdate, userID int64) (*serializers.ContactResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil || org.AccountID == nil {
		return nil, ErrOrganizationNotFound
	}

	updates := buildOrgContactUpdates(input, userID)
	var updateErr error
	if len(updates) > 0 {
		updateErr = s.orgRepo.UpdateContact(contactID, updates)
	}
	if updateErr != nil {
		return nil, updateErr
	}

	// For now return minimal response
	return &serializers.ContactResponse{ID: contactID}, nil
}

func buildOrgContactUpdates(input schemas.OrgContactUpdate, userID int64) map[string]interface{} {
	updates := map[string]interface{}{"updated_by": userID}
	if input.FirstName != nil {
		updates["first_name"] = *input.FirstName
	}
	if input.LastName != nil {
		updates["last_name"] = *input.LastName
	}
	if input.Email != nil {
		updates["email"] = *input.Email
	}
	if input.Phone != nil {
		updates["phone"] = *input.Phone
	}
	if input.Title != nil {
		updates["title"] = *input.Title
	}
	return updates
}

// DeleteOrganizationContact deletes a contact from an organization
func (s *OrganizationService) DeleteOrganizationContact(orgID int64, contactID int64) error {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return err
	}
	if org == nil || org.AccountID == nil {
		return ErrOrganizationNotFound
	}

	return s.orgRepo.DeleteContactFromAccount(*org.AccountID, contactID)
}

// GetOrganizationTechnologies returns technologies for an organization
func (s *OrganizationService) GetOrganizationTechnologies(orgID int64) ([]serializers.OrgTechnologyResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil {
		return nil, ErrOrganizationNotFound
	}

	techs, err := s.orgRepo.GetTechnologiesForOrg(orgID)
	if err != nil {
		return nil, err
	}

	results := make([]serializers.OrgTechnologyResponse, len(techs))
	for i, t := range techs {
		results[i] = serializers.OrgTechnologyResponse{
			OrganizationID: t.OrganizationID,
			TechnologyID:   t.TechnologyID,
			Source:         t.Source,
			DetectedAt:     t.DetectedAt.Format("2006-01-02T15:04:05Z"),
		}
		if t.Confidence != nil {
			f, _ := t.Confidence.Float64()
			results[i].Confidence = &f
		}
	}
	return results, nil
}

// AddTechnologyToOrganization adds a technology to an organization
func (s *OrganizationService) AddTechnologyToOrganization(orgID int64, input schemas.OrgTechnologyCreate) error {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return err
	}
	if org == nil {
		return ErrOrganizationNotFound
	}

	return s.orgRepo.AddTechnologyToOrg(orgID, input.TechnologyID, input.Source, input.Confidence)
}

// RemoveTechnologyFromOrganization removes a technology from an organization
func (s *OrganizationService) RemoveTechnologyFromOrganization(orgID int64, techID int64) error {
	return s.orgRepo.RemoveTechnologyFromOrg(orgID, techID)
}

// GetOrganizationFundingRounds returns funding rounds for an organization
func (s *OrganizationService) GetOrganizationFundingRounds(orgID int64) ([]serializers.FundingRoundResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil {
		return nil, ErrOrganizationNotFound
	}

	rounds, err := s.orgRepo.GetFundingRoundsForOrg(orgID)
	if err != nil {
		return nil, err
	}

	results := make([]serializers.FundingRoundResponse, len(rounds))
	for i, r := range rounds {
		results[i] = serializers.FundingRoundResponse{
			ID:             r.ID,
			OrganizationID: r.OrganizationID,
			RoundType:      r.RoundType,
			Amount:         r.Amount,
			LeadInvestor:   r.LeadInvestor,
			SourceURL:      r.SourceURL,
		}
		if r.AnnouncedDate != nil {
			d := r.AnnouncedDate.Format("2006-01-02")
			results[i].AnnouncedDate = &d
		}
	}
	return results, nil
}

// CreateOrganizationFundingRound creates a funding round for an organization
func (s *OrganizationService) CreateOrganizationFundingRound(orgID int64, input schemas.FundingRoundCreate) (*serializers.FundingRoundResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil {
		return nil, ErrOrganizationNotFound
	}

	round := &repositories.FundingRoundInput{
		OrganizationID: orgID,
		RoundType:      input.RoundType,
		Amount:         input.Amount,
		LeadInvestor:   input.LeadInvestor,
		SourceURL:      input.SourceURL,
		AnnouncedDate:  input.AnnouncedDate,
	}

	created, err := s.orgRepo.CreateFundingRoundFromInput(round)
	if err != nil {
		return nil, err
	}

	resp := &serializers.FundingRoundResponse{
		ID:             created.ID,
		OrganizationID: created.OrganizationID,
		RoundType:      created.RoundType,
		Amount:         created.Amount,
		LeadInvestor:   created.LeadInvestor,
		SourceURL:      created.SourceURL,
	}
	if created.AnnouncedDate != nil {
		d := created.AnnouncedDate.Format("2006-01-02")
		resp.AnnouncedDate = &d
	}
	return resp, nil
}

// DeleteOrganizationFundingRound deletes a funding round
func (s *OrganizationService) DeleteOrganizationFundingRound(orgID int64, roundID int64) error {
	return s.orgRepo.DeleteFundingRound(roundID)
}

// GetOrganizationSignals returns signals for an organization
func (s *OrganizationService) GetOrganizationSignals(orgID int64, signalType *string, limit int) ([]serializers.SignalResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil || org.AccountID == nil {
		return nil, ErrOrganizationNotFound
	}

	if limit <= 0 {
		limit = 20
	}

	signals, err := s.orgRepo.GetSignalsForAccount(*org.AccountID, signalType, limit)
	if err != nil {
		return nil, err
	}

	return serializers.SignalsToResponse(signals), nil
}

// CreateOrganizationSignal creates a signal for an organization
func (s *OrganizationService) CreateOrganizationSignal(orgID int64, input schemas.SignalCreate, userID int64) (*serializers.SignalResponse, error) {
	org, err := s.orgRepo.FindByID(orgID)
	if err != nil {
		return nil, err
	}
	if org == nil || org.AccountID == nil {
		return nil, ErrOrganizationNotFound
	}

	signalInput := repositories.SignalCreateInput{
		SignalType:  input.SignalType,
		Headline:    input.Headline,
		Description: input.Description,
		Source:      input.Source,
		SourceURL:   input.SourceURL,
		Sentiment:   input.Sentiment,
	}
	signalInput.SignalDate = lib.ParseDateString(input.SignalDate)
	if input.RelevanceScore != nil {
		dec := decimal.NewFromFloat(*input.RelevanceScore)
		signalInput.RelevanceScore = &dec
	}

	created, err := s.orgRepo.CreateSignalForAccount(*org.AccountID, signalInput)
	if err != nil {
		return nil, err
	}

	resp := serializers.SignalToResponse(created)
	return &resp, nil
}

// DeleteOrganizationSignal deletes a signal
func (s *OrganizationService) DeleteOrganizationSignal(orgID int64, signalID int64) error {
	return s.orgRepo.DeleteSignal(signalID)
}

// CreateOrganization creates a new organization
func (s *OrganizationService) CreateOrganization(input schemas.OrganizationCreate, userID int64, tenantID *int64) (*serializers.OrganizationDetailResponse, error) {
	// Use existing account if provided
	accountID := int64(0)
	if input.AccountID != nil {
		accountID = *input.AccountID
	}

	// Create new account if not provided
	if accountID == 0 {
		account, err := s.orgRepo.CreateAccount(input.Name, tenantID, userID)
		if err != nil {
			return nil, err
		}
		accountID = account.ID
	}

	org, err := s.orgRepo.CreateOrganization(repositories.OrganizationCreateInput{
		AccountID:            accountID,
		Name:                 input.Name,
		Website:              input.Website,
		Phone:                input.Phone,
		EmployeeCount:        input.EmployeeCount,
		EmployeeCountRangeID: input.EmployeeCountRangeID,
		FundingStageID:       input.FundingStageID,
		Description:          input.Description,
		RevenueRangeID:       input.RevenueRangeID,
		FoundingYear:         input.FoundingYear,
		HeadquartersCity:     input.HeadquartersCity,
		HeadquartersState:    input.HeadquartersState,
		HeadquartersCountry:  input.HeadquartersCountry,
		CompanyType:          input.CompanyType,
		LinkedInURL:          input.LinkedInURL,
		CrunchbaseURL:        input.CrunchbaseURL,
		UserID:               userID,
	})
	if err != nil {
		return nil, err
	}

	// Handle industry associations if provided
	if len(input.IndustryIDs) > 0 {
		s.orgRepo.UpdateAccountIndustries(accountID, input.IndustryIDs)
	}

	return s.GetOrganization(org.ID)
}
